Firestore/core/src/firebase/firestore/immutable/sorted_map_iterator.h

```diff
@@ -37,32 +37,36 @@ class SortedMapIterator {
   using difference_type = std::ptrdiff_t;
 
  public:
+  // Default constructor to conform to the requirements of ForwardIterator
+  SortedMapIterator() : tag_{Tag::Array}, array_iter_{} {
```

Comment:
Nit: I think initializing `array_iter{}` is redundant.

The only case I can think of when explicit value initialization is useful is when dealing with a C-style aggregate struct, because value-initializing class with no default constructor value-initializes (to zero, that is) all the members recursively. E.g.:
```cpp
struct Foo {
  int x;
  float y;
  char* z;
};
Foo bar{};
assert(bar.x == 0); // true
assert(bar.y == 0.f); // ~true
assert(bar.z == nullptr); // true
// A more verbose alternative
Foo baz;
baz.x = 0;
baz.y = 0.f;
baz.z = nullptr;
```

Reply:
Class members in anonymous unions that have nontrivial default constructors are not default initialized.

```c++
struct A {
  A() : value_(1) {}
  int value_;
};

struct B {
  B() : value_(2) {}
  int value_;
};

struct Foo {
  Foo() {}

  union {
    A a_;
    B b_;
  };
};

int main() {
  Foo foo;
  std::cout << foo.a_.value_ << std::endl;
}
```

When I compile and run this program the output is not `1`, which it would be if A's constructor were called. Changing `Foo`'s constructor to initialize the union with `Foo() : a_{} {}` makes the program work as expected.

See "Explanation" under http://en.cppreference.com/w/cpp/language/initializer_list

In this case, we should ensure that a postcondition of all constructors is that a variant in the union is chosen and initialized.

Reply:
Ah, I forgot about the union. Thanks for the explanation.